typing/printtyp.ml

```diff
@@ -2432,19 +2432,20 @@ let explain mis ppf =
   | None -> ()
   | Some explain -> explain ppf
 
-let warn_on_missing_def env ppf t =
+type env_kind = Normal_env | Abstract_env of string
+
+let warn_on_missing_def env ~env_kind ppf t =
   match get_desc t with
   | Tconstr (p,_,_) ->
     begin
       try
         ignore(Env.find_type p env : Types.type_declaration)
       with Not_found ->
-        match p with
-        | Path.Pident _ ->
+        match p, env_kind with
+        | Path.Pident _, Abstract_env where ->
```

Comment:
If this is not a `Pident`, the abstract environment is not the cause.
Of course, if we start to have other cases, this may not be the case, so in some way, your above suggestion of making the payload  a variant may make this code future-safe.